topic_menu: Add popover to view cross-topic links

evykassirer opened the pull request:
Fixes #35163.

<img width="302" height="378" alt="image" src="https://github.com/user-attachments/assets/7017027f-041d-4267-afa0-bd7cb6fe1100" />
<img width="460" height="167" alt="image" src="https://github.com/user-attachments/assets/035e27f8-18d7-4f44-b94a-a0601646dc00" />

evykassirer committed:
topic_menu: Add popover to view cross-link topics.

Fixes #35163.

evykassirer commented on web/src/message_store.ts:
```diff
@@ -394,16 +400,204 @@ export function reify_message_id({old_id, new_id}: {old_id: number; new_id: numb
+};
+
+const topic_links_by_from = new Map<number, Map<string, Set<TopicLink>>>();
+const topic_links_by_to = new Map<number, Map<string, Set<TopicLink>>>();
```
I went back and forth a lot on how to set up this data structure. This feels okay to me but still more involved than ideal. Let me know if you have better ideas!

evykassirer replied:
https://chat.zulip.org/#narrow/channel/6-frontend/topic/data.20structure.20for.20topic.20links/with/2284561

evykassirer commented on web/src/message_store.ts:
```diff
@@ -394,16 +400,204 @@ export function reify_message_id({old_id, new_id}: {old_id: number; new_id: numb
+    return new Set([...set].filter((item) => filter(item)));
+}
+
+function remove_topic_links(stream_id: number, topic: string, message_id: number): void {
```
I feel like this `remove_topic_links` and `save_topic_links` should maybe be async? I haven't done performance analysis on it yet and don't know how high priority the message edit/send is for being fast (especially since we have the local echo already)

timabbott replied:
If the data structures are done right, it should be O(1) to find the records to adjust. We should just avoid iterating through huge data structures.

evykassirer replied:
`remove_topic_links` is currently O(# of messages with links in the topic), which isn't ideal. But I initially tried storing them by message id and remembered not all links have an associated message id (e.g. links to topics and links to channels). But I just added back the map for storing by message id. I used `NO_MESSAGE_ID = 0` for when the `message_id` isn't set, does that seem reasonable?

evykassirer replied:
@timabbott does this feel resolved?

timabbott commented on web/src/topic_link_util.ts:
```diff
@@ -68,17 +68,17 @@ export function get_topic_link_content(
         if (message_id !== undefined) {
             return {
-                text: `#${escape(stream_name)} > ${escape(topic_display_name)} @ 💬`,
+                display_html: `#${escape(stream_name)} > ${escape(topic_display_name)} @ 💬`,
```
Hmm, I don't think this can be HTML; wouldn't `>` would need to be escaped if it's going to be that?

timabbott replied:
Posted a better suggestion for this below. Marking as resolved, we can manage via that thread.

timabbott commented on web/src/alert_words.ts:
```diff
@@ -56,30 +57,33 @@ export function process_message(message: Message): void {
-            },
+        message_store.update_message_content(
+            message,
+            message.content.replace(
```
Do `updated_content = message.content.replace(...`, the extra variable is helpful.

evykassirer commented on docs/THIRDPARTY:
```diff
@@ -333,6 +333,10 @@ Files: web/shared/icons/link-alt.svg
 Copyright: 2013-2023 Cole Bemis, Lucide Contributors 2025 (https://lucide.dev/icons/link)
 License: ISC License
 
+Files: web/shared/icons/link-2.svg
```
@alya do you want any input on naming this icon?

evykassirer replied:
To me they don't feel semantically different other than just being two different link images, and it's unclear to me when we would use one or the other. We could maybe call it `link-horizontal`?

<img width="168" height="54" alt="image" src="https://github.com/user-attachments/assets/e2400446-fd7c-4db0-92c6-b05563706b0e" />

evykassirer replied:
This ended up being added with "link-2" in 836b9dbafb261bb816ed589421ff642c5ccc3f70

evykassirer commented on web/src/message_store.ts:
```diff
@@ -389,9 +389,10 @@ export function update_status_emoji_info(
+        const message = message_data.message;
+        message.id = new_id;
+        message.locally_echoed = false;
+        stored_messages.set(new_id, {type: "server_message", message});
```
Yes, but a `LocalMessage` is just a `Message` with extra data, so the typing works fine here. We can explicitly remove those extra attributes?

But this commit isn't functionally changing anything (including types) so probably best discussed in #frontend?

evykassirer replied:
https://chat.zulip.org/#narrow/channel/6-frontend/topic/turning.20local.20message.20into.20server.20message

evykassirer replied:
Added to a PR here #36423

evykassirer commented on web/src/message_store.ts:
```diff
@@ -394,16 +400,204 @@ export function reify_message_id({old_id, new_id}: {old_id: number; new_id: numb
+    const stream_topic_regex =
+        /#narrow\/channel\/(\d+)-[^/]*(?:\/topic\/([^/]*)(?:\/near\/(\d+))?)?/;
+    for (const link of $(message.content).children("a")) {
+        const match = stream_topic_regex.exec(link.href);
```
I agree `find` would be better here than `children`

alya commented:
For header formatting, possibly a horizontal line and extra spacing too? Not sure. Somehow it looks a bit unfinished to me. We could also discuss in #design and pull in @terpimost for feedback.

evykassirer commented on web/src/stream_topic_history.ts:
```diff
@@ -362,6 +363,26 @@ export function get_latest_known_message_id_in_topic(
+// id of the topic in the cache, by encoding it at the end of the
+// traditional channel-topic url using a `with` operator. If client
+// cache doesn't have a message, we use the traditional link format.
+export function channel_topic_permalink_hash(stream_id: number, topic: string): string {
```
Turns out this wasn't necessary after some other refactors!

amanagr commented on web/src/topic_popover.ts:
```diff
@@ -54,186 +57,229 @@ function get_conversation(instance: tippy.Instance): {
+                // popover might run out of space when the second won't.
+                // I haven't found a way to get the real placement value other than
+                // getting from the DOM, so we're parsing it with zod.
+                // TODO(evy): figure out what's going on and do something less janky?
```
You can use `instance.reference._tippy.props.placement`

evykassirer replied:
No, that still has the value "bottom" even if the popover is actually flipping and displayed top.

evykassirer commented:
Next steps:

* Technical review: @amanagr helping with [this](https://github.com/zulip/zulip/pull/36293#discussion_r2479436830), people sharing thoughts on [sorting efficiency](https://github.com/zulip/zulip/pull/36293#discussion_r2479440168), and general thoughts on [#frontend > data structure for topic links](https://chat.zulip.org/#narrow/channel/6-frontend/topic/data.20structure.20for.20topic.20links/with/2286243)
* Fix unit tests. They're failing because of the use of JQuery for parsing links. If we decide we want to keep that, I can write tests, but I'm waiting to see if there are any comments about other nicer ways to extract the links.
* Design input and updates: header styling, header copy, new icons? [#design > cross-linking topics #35163 @ 💬](https://chat.zulip.org/#narrow/channel/101-design/topic/cross-linking.20topics.20.2335163/near/2299305)
* General product review (can happen concurrently or after the design things are complete)

timabbott commented on web/src/message_store.ts:
```diff
@@ -403,19 +409,361 @@ export function reify_message_id({old_id, new_id}: {old_id: number; new_id: numb
     message.content = new_content;
+    if (message.type !== "stream" || !may_have_new_links) {
+        return;
+    }
```
This block probably serves a comment on why DMs aren't included, possibly with a (not a priority) "TODO" note about what might make sense to do in DMs. (I think outgoing links from DM conversations could be useful to track, but would require new data structures?).

Also, is `may_have_new_links` the right check? I'd expect removing links to be relevant too...

evykassirer commented on web/src/message_store.ts:
```diff
@@ -403,19 +409,361 @@ export function reify_message_id({old_id, new_id}: {old_id: number; new_id: numb
+// Only stores the message id of the message linking to the narrow,
+// since we can get that message's narrow data from its id.
+// stream_id -> topic -> message_id -> set<from_message_id>
+const topic_links_by_to = new Map<number, Map<string, Map<number, number[]>>>();
```
I considered doing that, and I think it would be more readable, but `[stream_id, topic]` as an array key doesn't work because arrays with the same contents aren't equal to each other, and encoding the stream/topic as a string felt too fraught.

<img width="320" height="461" alt="image" src="https://github.com/user-attachments/assets/5b095d9d-23ef-4234-a510-77b8ffce7633" />

evykassirer commented:
Updates:

* Header styling: add icons and line under the header. Being discussed on CZO here: [#design > cross-linking topics #35163 @ 💬](https://chat.zulip.org/#narrow/channel/101-design/topic/cross-linking.20topics.20.2335163/near/2304068)
* Added comments for `sort_messages_by_timestamp`(https://github.com/zulip/zulip/pull/36293#discussion_r2540202212)
* Fixed `may_have_updated_links` name and added comment about DM links (https://github.com/zulip/zulip/pull/36293#discussion_r2525218428)
* Created a prep commit and unit tests for topic_util function, and opened prep PR with those: #36666
* Changed the quadratic code to linear, seeking feedback on the solution: https://github.com/zulip/zulip/pull/36293#discussion_r2540565369

timabbott commented on web/src/message_store.ts:
```diff
@@ -403,19 +409,361 @@ export function reify_message_id({old_id, new_id}: {old_id: number; new_id: numb
+// stream_id -> topic -> message_id -> set<from_message_id>
+const topic_links_by_to = new Map<number, Map<string, Map<number, number[]>>>();
+
+function sort_messages_by_timestamp(message_ids: number[]): number[] {
```
Well what I meant is you can sort by ID, not by `.timestamp`, not that you should assume the array is already sorted.

evykassirer replied:
Ah, so instead of using/calling this function we can sort the message ids. That makes sense and wasn't clear to me before. Thanks

evykassirer commented:
> Just noting that in the future, we'll have an option to restrict access to topics even when a user can access a channel (#19434), in case that's relevant to the implementation here.

At that point, we should probably add a check in the "If we don't have access to this stream (and/or message) or it's a buggy link, just ignore it" area and return early, but at this point there's no way to check that. Once this is merged, we could add it to that issue?

evykassirer commented:
Ah, new circular dependency is due to recently merged code I think. I'll look into it next time I'm working

evykassirer commented:
Rebased to remove conflicts

evykassirer commented:
Fixed conflicts

timabbott commented on web/src/topic_link_util.ts:
```diff
@@ -81,27 +83,38 @@ function _get_topic_link_content(opts: {
+    const text = escape_for_markdown ? `#${escape(stream_name)}` : `#${stream_name}`;
     return {
-        text: `#${escape(stream_name)}`,
+        text,
```
This new parameter seems unnecessary. I think you want this function to just return 3 values, replacing `text` with two values:
- `label_text_markdown`
- `label_text_plain`

In any case, this detail seems like a good opportunityy for a clean prep PR.

evykassirer commented on web/tests/example7.test.cjs:
```diff
@@ -74,7 +74,9 @@ const denmark_stream = make_stream({
 
-run_test("unread_ops", ({override}) => {
+run_test("unread_ops", ({override, override_rewire}) => {
+    override_rewire(message_store, "save_topic_links", noop);
```
There's jquery that happens in that function to get the links (I was talking about trouble testing this a while ago), which gives this error: `TypeError: Cannot read properties of undefined (reading 'selector'`

The alternative would be to do something like `set_message_links` in `message_store.test.cjs` but that's even more involved, and these tests aren't testing `set_message_links`, so I just mocked it out.

If you have a cleaner idea for what to do, I'd love to hear it!

evykassirer committed:
topic_popover: Refactor to allow different tippy placement directions.

evykassirer commented:
Trying a regex instead for getting the links from message content, to address [#design > cross-linking topics #36293 @ 💬](https://chat.zulip.org/#narrow/channel/101-design/topic/cross-linking.20topics.20.2336293/near/2379753) and also [addresses this review comment](https://github.com/zulip/zulip/pull/36293#discussion_r2710373729).

alexmv commented on web/src/message_store.ts:
```diff
@@ -722,18 +721,24 @@ function remove_message_from_topic_links(message_id: number): void {
-    // just for this feature, but there doesn't seem to be a nicer way to do it.
-    for (const link of [...$(message.content).find("a")].map((link) => link.href)) {
+    // Extract the URLs from the message content.
+    const links = /<a[^>]*?\s+href=(.*?)[\s,>]/.exec(message.content);
```
[One cannot parse HTML with regex](https://stackoverflow.com/questions/1732348/regex-match-open-tags-except-xhtml-self-contained-tags/1732454#1732454).

This is not my area of expertise, but I believe you want [`DOMParser().parseFromString(...)`](https://developer.mozilla.org/en-US/docs/Web/API/DOMParser/parseFromString).

evykassirer commented:
Rebased and resolved conflicts, and added Daniel's changes from https://github.com/evykassirer/zulip/pull/37 into the main commit

evykassirer commented:
Two updates from Aman's suggestions:

* Add a help page which helps user understand what links will be listed here and what each category means.
* Use our nice channel icon rendering template here

evykassirer commented:
Aman is re-reviewing this still on your radar? Probably would be good to test deploy again soon.

evykassirer commented:
Updates:
* Make the focus outline on the help icon :focus-visible so that it doesn't show up on click
* Remove links to channels
* Remove links to messages in the same topic

amanagr commented:
@evykassirer can you resolve the merge conflicts for me to deploy this?

evykassirer commented:
@amanagr done!

evykassirer commented:
Changes from comments on CZO:

```diff
diff --git a/starlight_help/src/content/docs/link-to-a-message-or-conversation.mdx b/starlight_help/src/content/docs/link-to-a-message-or-conversation.mdx
index 0bcb142cbb..cb5ec94295 100644
--- a/starlight_help/src/content/docs/link-to-a-message-or-conversation.mdx
+++ b/starlight_help/src/content/docs/link-to-a-message-or-conversation.mdx
@@ -250,8 +250,12 @@ to others through links, making it easy to navigate between related discussions.
 </Tabs>
 
 The **View links** option appears only when a conversation has links to or from
-it. These lists are built from messages Zulip has loaded, so they may be
-incomplete until you have visited the relevant conversations.
+it.
+
+<ZulipTip>
+  These lists are built from messages Zulip has loaded in your current session,
+  so they may be incomplete until you have visited the relevant conversations.
+</ZulipTip>
 
 ## Related articles
 
diff --git a/web/styles/popovers.css b/web/styles/popovers.css
index baaeb6bbd9..e8fe6d607b 100644
--- a/web/styles/popovers.css
+++ b/web/styles/popovers.css
@@ -1421,17 +1421,19 @@ ul.popover-menu-list {
     color: var(--color-text-popover-menu) !important;
 
     .popover-topic-links-list {
-        /* Equal padding on all sides; the bottom padding here stacks with
-           the next section title's top padding to leave a double-height
-           gap between the two link sections. */
-        padding: 0.6666em;
+        /* 3px at 15px/1em; with the separator's margin, this matches
+           the vertical rhythm of other popover menus. The larger bottom
+           padding stacks with the next section title's top padding to
+           leave a bigger gap between the two link sections. */
+        padding: 0.2em 0.6666em 0.6666em;
     }
 
     .topic-links-menu-title {
         display: flex;
         align-items: center;
         gap: 5px;
-        padding: 0.6666em;
+        /* 3px bottom at 15px/1em, matching other popover menus. */
+        padding: 0.6666em 0.6666em 0.2em;
         font-weight: 600;
         margin: 0;
 
@@ -1450,12 +1452,6 @@ ul.popover-menu-list {
         }
     }
 
-    .popover-menu-separator {
-        /* The space above and below each separator comes from the
-           adjacent title and list padding, so it stays symmetrical. */
-        margin: 0;
-    }
-
     .topic-links-menu-title,
     .popover-topic-links-list a {
         width: fit-content;
```

amanagr commented:
Thanks @evykassirer ! Merging since we are happy with the current deployed version on czo.

purepani commented:
Awesome that this got implemented!
